util/vec-map/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+/// heapless::Vec instead alloc::Vec.
+///
+/// It may be better to patch upstream for no_std compatibility and use that,
+/// but that crate has other issues -- it relies on an
+/// experimental "contracts" crate that causes a dependency on rand crate.
+/// Porting to Heapless would be a breaking change to the API.
+///
+/// TBD what the best path is: https://github.com/p-avital/vec-map-rs/blob/master/src/lib.rs
+#[derive(Clone, Debug)]
+pub struct VecMap<K, V, const N: usize> {
+    keys: Vec<K, N>,
+    values: Vec<V, N>,
+}
+
+impl<K, V, const N: usize> Default for VecMap<K, V, N> {
+    fn default() -> Self {
+        Self {
+            keys: Default::default(),
+            values: Default::default(),
+        }
+    }
+}
+
+impl<K, V, const N: usize> VecMap<K, V, N> {
+    /// Check if the map is empty
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get the number of items in the map
+    #[inline]
+    pub fn len(&self) -> usize {
+        debug_assert!(self.keys.len() == self.values.len());
+        self.keys.len()
+    }
+}
+
+impl<K: Eq + PartialEq, V, const N: usize> VecMap<K, V, N> {
+    /// Get the value associated to a key, if present
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.keys
+            .iter()
+            .position(|k| k == key)
+            .map(|idx| &self.values[idx])
+    }
+
+    /// Get a mutable reference to the value associated to a key, if present
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+        self.keys
+            .iter()
+            .position(|k| k == key)
+            .map(|idx| &mut self.values[idx])
+    }
+}
+
+impl<K: Clone + Eq + PartialEq, V, const N: usize> VecMap<K, V, N> {
+    /// Get a mutable reference to the value associated to a key, if present,
+    /// or else insert such a value produced by given callback,
+    /// and then return a mutable reference
+    ///
+    /// Returns an error if the heapless::Vec capacity was exceeded
+    pub fn get_mut_or_insert_with(
+        &mut self,
+        key: &K,
+        val_fn: impl FnOnce() -> V,
+    ) -> Result<&mut V, Error> {
+        if let Some(idx) = self.keys.iter().position(|k| k == key) {
+            Ok(&mut self.values[idx])
+        } else {
+            let idx = self.keys.len();
+            debug_assert_eq!(idx, self.values.len());
+            self.keys
+                .push(key.clone())
+                .map_err(|_| Error::CapacityExceeded)?;
+            self.values
+                .push(val_fn())
+                .map_err(|_| Error::CapacityExceeded)?;
+            Ok(&mut self.values[idx])
+        }
+    }
+
+    /// Get an iterator over the pairs in the VecMap
+    pub fn iter(&self) -> IterVecMap<K, V, N> {
+        IterVecMap::new(self)
+    }
+}
+
+// Sorting is possible when keys are ordered, and keys and values are cloneable
+impl<K: Clone + Ord + PartialOrd + Eq + PartialEq, V: Clone, const N: usize> VecMap<K, V, N> {
```

Comment:
`Ord` similarly extends `PartialOrd`

```suggestion
impl<K: Clone + Ord + Eq, V: Clone, const N: usize> VecMap<K, V, N> {
```
(throughout this file)